indexer/indexer.py

```diff
@@ -0,0 +1,337 @@
+    """Indexer class to help indexing operation.
+
+    The base class Indexer serves as the basis to extend other types
+    of transformative indexes. It codes a lot of the boiler plate code
+    regarding communication with ElasticSearch, so that users of this
+    base class only need to focus on their indexing algorithm. Users
+    may optionally overwrite all of the methods should they need to use
+    other technologies besides ElasticSearch.
+
+    The index() method is the main method for the Indexer class. Users
+    should ideally only overwrite index(), special_fields() and merge().
+    index() should call special_fields() first and then merge(). Users
+    can define their own extra functions to help with indexing as needed.
+
+    """
+
+    def __init__(self, metadata_files, data_files, es_client, index_name,
+                 doc_type, index_settings=None, index_mapping_config=None,
+                 **kwargs):
+        """
+        Create an instance of the Indexing class.
+
+        The constructor creates an Indexer instance, with various
+        parameters that are needed when performing the indexing operation.
+        It requires a dictionary of the metadata_files, the data_files,
+        an ElasticSearch(ES) client object to communicate with some
+        ES instance, the index_settings, as well as a mapping config
+        file that can be used by the indexer for indexing operation and
+        actual ElasticSearch mapping. Any extra kwargs are set as
+        attributes to the instance.
+
+        :param metadata_files: dictionary of json metadata (list of dicts)
+        :param data_files: dictionary describing non-metadata files
+         (list of dicts)
+        :param es_client: The elasticsearch client
+        :param index_name: The name of the index to put the documents.
+        :param doc_type: The document type to put the document under in ES.
+        :param index_settings: Any special settings for the ES index.
+        :param index_mapping_config: The indexer config file.
+        """
+        # Set main arguments
+        self.metadata_files = metadata_files
+        self.data_files = data_files
+        self.es_client = es_client
+        self.index_name = index_name
+        self.doc_type = doc_type
+        self.index_settings = index_settings
+        self.index_mapping_config = index_mapping_config
+        # Set all kwargs as instance attributes
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+
+    def index(self, bundle_uuid, bundle_version, *args, **kwargs):
+        """
+        Indexes the data files.
+
+        Triggers the actual indexing process
+
+        :param bundle_uuid: The bundle_uuid of the bundle that will be indexed.
+        :param bundle_version: The bundle of the version.
+        """
+        raise NotImplementedError(
+            'users must define index to use this base class')
+
+    def special_fields(self, *args, **kwargs):
```

Comment:
Eliminate any `*args` and `**kwargs` unless you really need them. Don't put those in as an attempt to make future changes easier. They don't.